frame/tx-pause/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Pause a call.
+		///
+		/// Can only be called by [`Config::PauseOrigin`].
+		/// Emits an [`Event::CallPaused`] event on success.
+		#[pallet::weight(0)]
+		pub fn pause_call(
+			origin: OriginFor<T>,
+			pallet: PalletNameOf<T>,
+			extrinsic: ExtrinsicNameOf<T>,
+		) -> DispatchResult {
+			T::PauseOrigin::ensure_origin(origin)?;
+
+			Self::ensure_can_pause(&pallet, &extrinsic)?;
+			PausedCalls::<T>::insert((&pallet, &extrinsic), ());
+			Self::deposit_event(Event::CallPaused(pallet, extrinsic));
+
+			Ok(())
+		}
+
+		/// Un-pause a call.
+		///
+		/// Can only be called by [`Config::UnpauseOrigin`].
+		/// Emits an [`Event::CallUnpaused`] event on success.
+		#[pallet::weight(0)]
+		pub fn unpause_call(
+			origin: OriginFor<T>,
+			pallet: PalletNameOf<T>,
+			extrinsic: ExtrinsicNameOf<T>,
+		) -> DispatchResult {
+			T::UnpauseOrigin::ensure_origin(origin)?;
+
+			Self::ensure_can_unpause(&pallet, &extrinsic)?;
+			PausedCalls::<T>::remove((&pallet, &extrinsic));
+			Self::deposit_event(Event::CallUnpaused(pallet, extrinsic));
+
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Return whether this call is paused.
+	pub fn is_paused_unbound(pallet: Vec<u8>, extrinsic: Vec<u8>) -> bool {
+		let pallet = PalletNameOf::<T>::try_from(pallet);
+		let extrinsic = ExtrinsicNameOf::<T>::try_from(extrinsic);
+
+		match (pallet, extrinsic) {
+			(Ok(pallet), Ok(extrinsic)) => Self::is_paused(&pallet, &extrinsic),
+			_ => T::PauseTooLongNames::get(),
+		}
+	}
+
+	/// Return whether this call is paused.
+	pub fn is_paused(pallet: &PalletNameOf<T>, extrinsic: &ExtrinsicNameOf<T>) -> bool {
+		<PausedCalls<T>>::contains_key((pallet, extrinsic))
+	}
+
+	/// Ensure that this call can be paused.
+	pub fn ensure_can_pause(
+		pallet: &PalletNameOf<T>,
+		extrinsic: &ExtrinsicNameOf<T>,
+	) -> Result<(), Error<T>> {
+		// The `TxPause` pallet can never be paused.
+		if pallet.as_ref() == <Self as PalletInfoAccess>::name().as_bytes().to_vec() {
+			return Err(Error::<T>::IsUnpausable)
+		}
+		if T::UnpausablePallets::contains(&pallet) {
+			return Err(Error::<T>::IsUnpausable)
+		}
+		if Self::is_paused(pallet, extrinsic) {
+			return Err(Error::<T>::IsPaused)
+		}
+		Ok(())
+	}
+
+	/// Ensure that this call can be un-paused.
+	pub fn ensure_can_unpause(
+		pallet: &PalletNameOf<T>,
+		extrinsic: &ExtrinsicNameOf<T>,
+	) -> Result<(), Error<T>> {
+		if Self::is_paused(pallet, extrinsic) {
+			// SAFETY: Everything that is paused, can be un-paused.
+			Ok(())
+		} else {
+			Err(Error::IsUnpaused)
+		}
+	}
+}
+
+impl<T: pallet::Config> Contains<T::Call> for Pallet<T>
+where
+	<T as frame_system::Config>::Call: GetCallMetadata,
+{
+	/// Return whether the call is allowed to be dispatched.
+	fn contains(call: &T::Call) -> bool {
+		let CallMetadata { pallet_name, function_name } = call.get_call_metadata();
+		!Pallet::<T>::is_paused_unbound(pallet_name.into(), function_name.into())
+	}
```

Comment:
very interesting

Reply:
I went this way instead of passing through a default filter since filters can be composed liked this:  
```rust
InsideBoth<SafeMode, TxPause>
```